server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java

```diff
@@ -342,15 +341,20 @@ public void canReturnNullResponseIfMatchNoDocs(boolean value) {
         this.canReturnNullResponseIfMatchNoDocs = value;
     }
 
+    private static final ThreadLocal<BytesStreamOutput> scratch = ThreadLocal.withInitial(BytesStreamOutput::new);
+
     /**
      * Returns the cache key for this shard search request, based on its content
      */
     public BytesReference cacheKey() throws IOException {
-        BytesStreamOutput out = new BytesStreamOutput();
```

Comment:
This was really inefficient before. The cache key would usually be ~100b but be written with buffers of sizes `8`, `56` and then `1xx`, then copied again below.